### ISSUE TITLE
App crash with deleteDocument with concurrencyControl

### ISSUE THREAD

**Issue by sridevi-15** [couchbase/couchbase-lite-android#1652]
### Steps to reproduce : 
1. Create document id = doc1 as doc1a instance
    2. Get the cbl doc and update the doc with concurrency
    3. delete same doc with concurrency
    4. Verify  (success for LastWriteWins, fail for FailOnConflict)

### Result : 
1. app crashes

CBL logs : https://gist.github.com/sridevi-15/6203e5579a9516bde547e1258b77739e

### Environment : 
sg -> 2.0.0 821
cbl Android . -> 2.0.0-401 ee

###  Code : 
https://github.com/couchbaselabs/mobile-testkit/blob/feature/cbl20-ver1/testsuites/CBLTester/CBL_Functional_tests/test_concurrency.py#L92

### command to run : 
pytest -rsx -s -k test_replication_with_concurrencyControl_deleteSameDocId --server-version=5.0.0 --sync-gateway-version=2.0.0-821 --mode=cc --liteserv-host=192.168.33.101 --liteserv-port=8080 --xattrs --liteserv-version=2.0.0-401 --liteserv-platform=android --create-db-per-test=cbl-test --skip-provisioning testsuites/CBLTester/CBL_Functional_tests/

**Comment by hideki** [comment]
Hi @sridevi-15 

I find only one line that tell crash. Can you capture entire full stack trace?
`03-07 12:55:35.685 12887-12912/com.couchbase.TestServerApp A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x0 in tid 12912 (NanoHttpd Reque)`

**Comment by sridevi-15** [comment]
https://gist.github.com/sridevi-15/d291cd84897fe244d90ad166d60018aa

New cbl logs : https://gist.github.com/sridevi-15/962f0cb55e7f80a0b2b40767da7fa89b

**Comment by hideki** [comment]
@sridevi-15 do you have python log? I'd like to identify which operation fails

**Comment by hideki** [comment]
Note to self:

The root cause of the crash:
```
        // 2. update
        Document doc = db.getDocument("doc");
        mdoc = doc.toMutable();
        mdoc.setString("update", "updated");
        db.save(mdoc);

        // 3. delete by previously retrived document
        assertTrue(db.delete(doc, ConcurrencyControl.LAST_WRITE_WINS));
```
1. ` mdoc = doc.toMutable();` -> `mdoc` and `doc` points to same instance of `C4Document`
2.  `db.save(mdoc);` -> `mdoc` has new instance of `C4Document`. old instance of `C4Document` is freed. `doc` points to `C4Document` java instance which points to `NULL` of C4Document native.
3. `db.delete(doc, ConcurrencyControl.LAST_WRITE_WINS)` access `C4Document` native instance. But it is already freed.

Possible solution
- Introduce reference counting to C4Document java class

**Comment by hideki** [comment]
Please verify the fix with 2.0.0-429 (build in progress) or higher
http://mobile.maven.couchbase.com/feeds/dev//com/couchbase/lite/couchbase-lite-android-ee/2.0.0-429

**Comment by sridevi-15** [comment]
verified on build 429 , it worked
